sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala

```diff
@@ -1044,6 +1044,34 @@ class Analyzer(
           failOnOuterReference(p)
           p
       }
+
+      // SPARK-17348
+      // Looking for a potential incorrect result case.
+      // When a correlated predicate is a non-equality predicate
+      // it must be placed at the immediate child operator.
+      // Otherwise, the pull up of the correlated predicate
+      // will generate a plan with a different semantics
+      // which could return incorrect result.
+      var continue : Boolean = true
```

Comment:
@hvanhovell I totally agree with you on sharing different perspectives and I don't mind to share what I can. I would like to draw up a plan for subquery processing but I hesitate to commit to a full support in 2.2. When is the rough date for 2.2 code cut-off? Given that I am relatively new to Spark code base, I would seek the commitment from someone like you (or any committers) who has experience in Spark SQL to review my design proposal and code. You would be an ideal candidate since you implemented most, if not all, the subquery functionality in 2.0.

On the second part of your comment, I do not know the code that does the semantics resolution from a unresolved LogicalPlan to a fully resolved LogicalPlan. I make this comment based on my general knowledge of SQL compilation. I lean toward to agree with your last statement that a complete isolation between resolving a reference (current or outer) and the rewrite. I believe this is an important principle we want to keep in the design of a SQL compiler and optimizer. How to do that cleanly? The above SQL is equivalent to  
````
select ... 
from (
      select b.key, min(b.value) as min_value
      from src b
      group by b.key) as Q
where exists (select a.key
              from src a
              where a.value > 'val_9' and a.value = Q.min_value)
````

To generalize further, the expression(s) referencing the outer reference, B in this case, in the HAVING clause needs not to be the same as in the SELECT clause, i.e., the aggregate function in the EXISTS subquery can be any aggregate function on any column in B. Spark supports this today.

What we may need is to recognize that there is an expression referencing outer reference in the subquery and that expression, if does not exist, need to be added to the projection list of the outer reference and the identifier of that expression will be used to substitute the whole expression in the subquery.

I'd like to propose we spin off the discussion of this topic in a separate thread, both on the technical detail and the schedule.